Awilliams/fix basemapgallery set incorrect sp

mrwillyees opened the pull request:
Fix to avoid the early set basemap also for the items with a not matching spatial reference. 
Wait the first baselayer to be loaded and then check the sp. If not matching, don't set the basemap and set the item as disabled by forcing the listview to redraw the items.

anmacdonald commented on uitools/cpp/Esri/ArcGISRuntime/Toolkit/BasemapGalleryController.cpp:
```diff
@@ -508,16 +508,38 @@ namespace Toolkit {
+        if(e.isEmpty()){
+            if (basemap == m_currentBasemap)
+                  return;
+            if(!basemapMatchesCurrentSpatialReference(basemap)){
```
🙃  Has the formatter been run on this? Our coding conventions are always a newline between the if-statement and the brace.

anmacdonald commented on uitools/cpp/Esri/ArcGISRuntime/Toolkit/BasemapGalleryController.cpp:
```diff
@@ -508,16 +508,38 @@ namespace Toolkit {
+
+      auto apply = [basemap, this, connection](Error e){
+        if(e.isEmpty()){
+            if (basemap == m_currentBasemap)
+                  return;
+            if(!basemapMatchesCurrentSpatialReference(basemap)){
+                //force redraw for all the lsitview items.
+                emit m_gallery->dataChanged(m_gallery->index(0), m_gallery->index(std::max(m_gallery->rowCount() - 1, 0)));
+                return;
```
❌  I'd maybe put `connection` in a `shared_ptr`. There's memory leaks in these early returns.

anmacdonald commented on uitools/cpp/Esri/ArcGISRuntime/Toolkit/BasemapGalleryController.cpp:
```diff
@@ -508,16 +508,38 @@ namespace Toolkit {
+        delete connection;
+      };
+      if(basemap->baseLayers()->first()->loadStatus() != LoadStatus::Loaded){
+          connect(basemap->baseLayers()->first(), &Layer::doneLoading, this, apply);
```
❌  Missing `*connection = connect(`

anmacdonald commented on uitools/cpp/Esri/ArcGISRuntime/Toolkit/BasemapGalleryController.cpp:
```diff
@@ -611,11 +633,13 @@ namespace Toolkit {
-                         const auto sp2 = layer->spatialReference();
-                         return sp2.isEmpty() || sp == sp2;
-                       });
+    if(layers->size() > 0){
```
🙃 Brace on newline.

anmacdonald commented on uitools/cpp/Esri/ArcGISRuntime/Toolkit/BasemapGalleryController.cpp:
```diff
@@ -508,16 +508,38 @@ namespace Toolkit {
+          basemap->baseLayers()->first()->load();
+      }
+      else {
+          apply(Error());
```
💭  I'd maybe go for `Error{}` instead. Does this work on all compilers? Looks like a potential most-vexing parse problem.

anmacdonald reviewed (changes requested):
Looks good. A couple of potential issues in connection managemen and formatting.

mrwillyees commented on uitools/cpp/Esri/ArcGISRuntime/Toolkit/BasemapGalleryController.cpp:
```diff
@@ -607,15 +641,46 @@ namespace Toolkit {
+    if(layers->size() <= 0)
+      return false;
+
+    if(auto sceneView = static_cast<Scene*>(m_geoModel))
```
💭 Should it not actually build if it is not able to cast to the specified type when using the static_cast? How can get a segmentation fault?

anmacdonald replied:
We have

```
class Map: public GeoModel
```

and

```
class Scene: public GeoModel
```

The given GeoModel can be either a Map or a Scene. Here we are casting the GeoModel to a Scene without any test to see if it might be a Map instead. `static_cast` is a compile-time check only, not a runtime check, and casting a GeoModel to a Scene is always a valid type conversion regardless of the underlying data.

anmacdonald replied:
This might help illustrate https://onlinegdb.com/mQheT-BDB

mrwillyees commented on uitools/cpp/Esri/ArcGISRuntime/Toolkit/BasemapGalleryController.cpp:
```diff
@@ -607,15 +641,46 @@ namespace Toolkit {
+
+    const auto layer = layers->first();
+    const auto sp2 = layer->spatialReference();
+    qDebug() << "sp1 " << sp.toJson();
+    qDebug() << "sp2 "<< sp2.toJson();
```
yea, should start properly check for all qDebugs before commiting

anmacdonald reviewed (approved):
👍 LGTM.

Do doubl-check you have installed the clangformat plugin in Qt Creator, as I am still seeing inconsistent brace usage.

mrwillyees commented:
After last review: added first layer check for qml controller. Also added scene checks using the sceneViewTilingScheme. A busy spinner used to visually hint to the user that the basemap is being loaded

anmacdonald commented on uitools/import/Esri/ArcGISRuntime/Toolkit/BasemapGallery.qml:
```diff
@@ -199,17 +199,28 @@ Pane {
+
+            Connections {
+                target: controller
+                function onCurrentBasemapChanged() { busyIndicator.running = false;}
```
🙃 
```suggestion
                function onCurrentBasemapChanged() { 
                    busyIndicator.running = false;
                }
```

anmacdonald commented on uitools/import/Esri/ArcGISRuntime/Toolkit/BasemapGallery.qml:
```diff
@@ -228,16 +239,21 @@ Pane {
                 hoverEnabled: true
-                onClicked: controller.setCurrentBasemap(modelData.basemap)
+                onClicked: {
+                    if(controller.currentBasemap !== modelData.basemap)
```
🙃 
```suggestion
                    if (controller.currentBasemap !== modelData.basemap)
```

anmacdonald commented on uitools/import/Esri/ArcGISRuntime/Toolkit/BasemapGallery.qml:
```diff
@@ -228,16 +239,21 @@ Pane {
                     timerOnEntered.showTooltipFn = () => {
                         if (allowTooltips && mouseArea.containsMouse && modelData.tooltip !== "")
-                            basemapDelegate.ToolTip.visible = true;
+                        basemapDelegate.ToolTip.visible = true;
```
🙃 
```suggestion
                            basemapDelegate.ToolTip.visible = true;
```

anmacdonald commented on uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BasemapGalleryController.qml:
```diff
@@ -81,8 +81,16 @@ QtObject {
+        //set connection with matchescurrentsp to run after the basemap is loaded
+        const layer = basemap.baseLayers.get(0);
+        layer.loadStatusChanged.connect(internal.setCurrentBasemapMatchingSp.bind(null, basemap, layer));
+        if(layer.loadStatus !== Enums.LoadStatusLoaded) {
```
🙃 
```suggestion
        if (layer.loadStatus !== Enums.LoadStatusLoaded) {
```

anmacdonald commented on uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BasemapGalleryController.qml:
```diff
@@ -81,8 +81,16 @@ QtObject {
+        layer.loadStatusChanged.connect(internal.setCurrentBasemapMatchingSp.bind(null, basemap, layer));
+        if(layer.loadStatus !== Enums.LoadStatusLoaded) {
+            layer.load();
+        }
+        else {
```
🙃 
```suggestion
        } else {
```

anmacdonald commented on uitools/import/Esri/ArcGISRuntime/Toolkit/BasemapGallery.qml:
```diff
@@ -228,16 +239,21 @@ Pane {
             ToolTip.text: modelData.tooltip
             MouseArea {
                 id: mouseArea
+                z : 2
```
❓ What does this do?

mrwillyees replied:
I set the mouse area above everything else, so when a click happens it is recorded by the area and not from the Image. Otherwise only the bottom part of the item is clickable

anmacdonald replied:
Does this still happen with the change to `ItemDelegate`?

mrwillyees replied:
Yes, but is because I have to set the [`z`](https://github.com/Esri/arcgis-runtime-toolkit-qt/pull/462/files#diff-ca77a4aefb9091361317a2c42ed3a2b69aaafbb1a6ae456189aec5cf38814285R202) property of the indicator object > 0, otherwise is rendered behind the `Image`, but doing so, the indicator will be over the `MouseArea`

https://user-images.githubusercontent.com/91129623/142215849-c785ca96-279a-4301-8824-a331ae83b573.mp4

.

anmacdonald replied:
I see, that makes sense. Good find and good fix!

anmacdonald commented on uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BasemapGalleryController.qml:
```diff
@@ -169,28 +177,47 @@ QtObject {
-            if (layer === null) {
-                continue;
+        let item = basemap.item;
+        if(item !== null){
```
🙃 
```suggestion
        if (item !== null) {
```

anmacdonald commented on uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BasemapGalleryController.qml:
```diff
@@ -169,28 +177,47 @@ QtObject {
+        let item = basemap.item;
+        if(item !== null){
+            let it_sp = item.spatialReference;
+            if(item !== null && it_sp !== null){
```
🙃 
```suggestion
            if (item !== null && it_sp !== null) {
```

anmacdonald commented on uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BasemapGalleryController.qml:
```diff
@@ -169,28 +177,47 @@ QtObject {
-            }
+        let layers = basemap.baseLayers;
+
+        if(layers.count <= 0)
```
🙃 
```suggestion
        if (layers.count <= 0)
```

anmacdonald commented on uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BasemapGalleryController.qml:
```diff
@@ -169,28 +177,47 @@ QtObject {
 
-            if (!sp2.equals(sp1)) {
-                return false;
+        //scene case
```
🙃 
```suggestion
        // scene case
```

anmacdonald commented on uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BasemapGalleryController.qml:
```diff
@@ -169,28 +177,47 @@ QtObject {
         }
-        return true;
+
+        //map case
```
🙃 
```suggestion
        // map case
```

anmacdonald commented on uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BasemapGalleryController.qml:
```diff
@@ -169,28 +177,47 @@ QtObject {
+            if(sp2 === null)
+                return true;
+            let svts = geoModel.sceneViewTilingScheme;
+            switch(svts){
```
🙃 
```suggestion
            switch(svts) {
```

anmacdonald commented on uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BasemapGalleryController.qml:
```diff
@@ -169,28 +177,47 @@ QtObject {
-            if (!sp2.equals(sp1)) {
-                return false;
+        //scene case
+        if(geoModel instanceof Scene){
```
🙃 
```suggestion
        if (geoModel instanceof Scene) {
```

anmacdonald commented on uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BasemapGalleryController.qml:
```diff
@@ -169,28 +177,47 @@ QtObject {
+        if(geoModel instanceof Scene){
+            let layer = layers.get(0);
+            let sp2 = layer.spatialReference;
+            if(sp2 === null)
```
🙃 
```suggestion
            if (sp2 === null)
```

anmacdonald commented on uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BasemapGalleryController.qml:
```diff
@@ -221,6 +248,21 @@ QtObject {
         }
 
+        function setCurrentBasemapMatchingSp(basemap, layer) {
+            if(layer.loadStatus === Enums.LoadStatusLoaded) {
```
🙃 
```suggestion
            if (layer.loadStatus === Enums.LoadStatusLoaded) {
```

anmacdonald commented on uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BasemapGalleryController.qml:
```diff
@@ -221,6 +248,21 @@ QtObject {
+        function setCurrentBasemapMatchingSp(basemap, layer) {
+            if(layer.loadStatus === Enums.LoadStatusLoaded) {
+                if(!basemapMatchesCurrentSpatialReference(basemap)){
+                    //manually setting the currbasemap to trigger the event onCurrentBasemapChaged
```
🙃 
```suggestion
                    // manually setting the currbasemap to trigger the event onCurrentBasemapChaged
```

anmacdonald commented on uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BasemapGalleryController.qml:
```diff
@@ -221,6 +248,21 @@ QtObject {
+                    internal.gallery.dataChanged(internal.gallery.index(basemapIndex(basemap), 0), internal.gallery.index(basemapIndex(basemap), 0));
+                    return;
+                }
+                internal.currentBasemap = basemap;
+                geoModel.basemap = internal.currentBasemap;
```
❓ If the geoModel basemap is being set, does the `internal.currentBasemap` need set? I'm hoping it would update automatically.

mrwillyees replied:
❓ Not sure what you mean exactly. If we don't set the `internal.currentBasemap`, then the `geoModel.basemap` will be set to the old `internal.currentBasemap`. Did you mean the other way around, to not set `geoModel.basemap` manually?

anmacdonald replied:
There's a connections object at line 243 which I'd hope would keep the two in sync, as the geomodel basemap can be changed from outwith the gallery anyway.

mrwillyees replied:
❗ found out that self assignment will not trigger the binding, so 
`geoModel.basemap = geoModel.basemap` will not set the `internal.currentBasemap`. This is shown by the busy indicator keep spinning for the New Zealand basemap, which is stopped upon the raise of onCurrentBasemapChanged()

anmacdonald commented on uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/BasemapGalleryController.qml:
```diff
@@ -221,6 +248,21 @@ QtObject {
+                }
+                internal.currentBasemap = basemap;
+                geoModel.basemap = internal.currentBasemap;
+                layer.loadStatusChanged.disconnect(setCurrentBasemapMatchingSp);
```
🔧 An early return may prevent this disconnet from being called, I suggest moving this to the begining of the function like so:

```qml
if (layer.loadStatus === Enums.LoadStatusLoaded) {
    layer.loadStatusChanged.disconnect(setCurrentBasemapMatchingSp);
    ....
```

I used to work for a company that disallowed return statements as part of control-flow. While this is not a requirement at Esri, I would encourage being sparse with this style so that these kinds of bugs can be avoided.

anmacdonald reviewed (approved):
👍 LGTM. Couple of questions/suggestions. Some styling pedantry on my end.

No need to re-review after you've applied my changes, all looks good! Thanks for doing this.
